File: strategy/typescript/base/coordinates.ts
/**
 * @module coordinates
 * Functions to convert from global to strategy let coordinates and back.
 * Only use to convert values from or for amun!
 */

/**************************************************************************
*   Copyright 2018 Alexander Danzer, Michael Eischer, Andreas Wendler     *
*   Robotics Erlangen e.V.                                                *
*   http://www.robotics-erlangen.de/                                      *
*   info@robotics-erlangen.de                                             *
*                                                                         *
*   This program is free software: you can redistribute it and/or modify  *
*   it under the terms of the GNU General Public License as published by  *
*   the Free Software Foundation, either version 3 of the License, ||     *
*   any later version.                                                    *
*                                                                         *
*   This program is distributed in the hope that it will be useful,       *
*   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
*   MERCHANTABILITY or  FITNESS FOR A PARTICULAR PURPOSE.  See the        *
*   GNU General Public License for more details.                          *
*                                                                         *
*   You should have received a copy of the GNU General Public License     *
*   along with this program.  If not, see <http://www.gnu.org/licenses/>. *
**************************************************************************/


import { Vector } from "base/vector";

interface CoordinatesType {
	/**
	 * Converts strategy local coordinates to global coordinates for amun
	 * @param data - vector to convert
	 */
	toGlobal(pos: Vector): Vector;
	/**
	 * Converts strategy local coordinates to global coordinates for amun
	 * @param data - vector to convert
	 */
	toGlobal(pos: Readonly<Vector>): Readonly<Vector>;
	/**
	 * Converts strategy local coordinates to global coordinates for amun
	 * @param num - angle to convert
	 */
	toGlobal(num: number): number;
	/**
	 * Converts global coordinates from amun to strategy let coordinates
	 * @param pos - vector to convert
	 */
	toLocal(pos: Vector): Vector;
	/**
	 * Converts global coordinates from amun to strategy let coordinates
	 * @param pos - vector to convert
	 */
	toLocal(pos: Readonly<Vector>): Readonly<Vector>;
	/**
	 * Converts global coordinates from amun to strategy let coordinates
	 * @param num - angle to convert
	 */
	toLocal(num: number): number;
	/**
	 * Does toGlobal conversion for a list
	 * @param data - list to map
	 */
	listToGlobal(pos: Vector[]): Vector[];
	/**
	 * Does toGlobal conversion for a list
	 * @param data - list to map
	 */
	listToGlobal(pos: Readonly<Vector>[]): Readonly<Vector>[];
	/**
	 * Does toGlobal conversion for a list
	 * @param data - list to map
	 */
	listToGlobal(num: number[]): number[];
	/**
	 * Converts strategy local coordinates to vision coordinates for amun
	 * @param data - vector to convert
	 */
	toVision(pos: Vector): Vector;
	/**
	 * Converts strategy local coordinates to vision coordinates for amun
	 * @param data - vector to convert
	 */
	toVision(pos: Readonly<Vector>): Readonly<Vector>;
	/**
	 * Converts strategy local coordinates to vision coordinates for amun
	 * @param data - angle to convert
	 */
	toVision(num: number): number;
	/**
	 * Converts vision coordinates to strategy local coordinates
	 * @param data - vector to convert
	 */
	fromVision(pos: Vector): Vector;
	/**
	 * Converts vision coordinates to strategy local coordinates
	 * @param data - vector to convert
	 */
	fromVision(pos: Readonly<Vector>): Readonly<Vector>;
	/**
	 * Converts vision coordinates to strategy local coordinates
	 * @param data - angle to convert
	 */
	fromVision(pos: number): number;
}

class Invert implements CoordinatesType {
	toGlobal(data: any): any {
		if (typeof(data) === "number") {
			let num = data as number;
			if (num > Math.PI) {
				return num - Math.PI;
			} else {
				return num + Math.PI;
			}
		} else {
			let vector = data as Vector;
			return new Vector(-vector.x, -vector.y);
		}
	}
	toLocal(data: any): any {
		return this.toGlobal(data);
	}
	listToGlobal(data: any[]): any[] {
		let inverted = [];
		for (let v of data) {
			inverted.push(this.toGlobal(v));
		}
		return inverted;
	}
	toVision(data: any): any {
		if (typeof(data) === "number") {
			// rotate 270 degrees
			let num = data as number - Math.PI * 1.5;
			if (num < 0) {
				num += 2.0 * Math.PI;
			}
			return num;
		} else {
			// meter to millimeter
			let vector = data as Vector * 1000;
			return new Vector(-vector.y, vector.x);
		}
	}
	fromVision(data: any): any {
		if (typeof(data) === "number") {
			// rotate 270 degrees
			let num = data as number + Math.PI * 1.5;
			if (num > 2.0 * Math.PI) {
				num -= 2.0 * Math.PI;
			}
			return num;
		} else {
			// meter to millimeter
			let vector = data as Vector * 0.001;
			return new Vector(vector.y, -vector.x);
		}
	}
}

class Pass implements CoordinatesType {
	toGlobal(value: any): any {
		return value;
	}
	toLocal(value: any): any {
		return value;
	}
	listToGlobal(value: any): any {
		return value;
	}
	toVision(data: any): any {
		if (typeof(data) === "number") {
			// rotate 90 degrees
			let num = data as number - Math.PI * 0.5;
			if (num < 0) {
				num += 2.0 * Math.PI;
			}
			return num;
		} else {
			// meter to millimeter
			let vector = data as Vector * 1000;
			return new Vector(vector.y, -vector.x);
		}
	}
	fromVision(data: any): any {
		if (typeof(data) === "number") {
			// rotate 90 degrees
			let num = data as number + Math.PI * 0.5;
			if (num > 2.0 * Math.PI) {
				num -= 2.0 * Math.PI;
			}
			return num;
		} else {
			// meter to millimeter
			let vector = data as Vector * 0.001;
			return new Vector(-vector.y, vector.x);
		}
	}
}

// eslint-disable-next-line @typescript-eslint/naming-convention
export let Coordinates: CoordinatesType;

export function _setIsBlue(teamIsBlue: boolean) {
	if (teamIsBlue) {
		Coordinates = new Invert();
	} else {
		Coordinates = new Pass();
	}
}

File: strategy/lua/base/coordinates.lua
--[[
--- Functions to convert from global to strategy local coordinates and back.
-- Only use to convert values from or for amun!
module "Coordinates"
]]--

--[[***********************************************************************
*   Copyright 2015 Alexander Danzer, Michael Eischer                      *
*   Robotics Erlangen e.V.                                                *
*   http://www.robotics-erlangen.de/                                      *
*   info@robotics-erlangen.de                                             *
*                                                                         *
*   This program is free software: you can redistribute it and/or modify  *
*   it under the terms of the GNU General Public License as published by  *
*   the Free Software Foundation, either version 3 of the License, or     *
*   any later version.                                                    *
*                                                                         *
*   This program is distributed in the hope that it will be useful,       *
*   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
*   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
*   GNU General Public License for more details.                          *
*                                                                         *
*   You should have received a copy of the GNU General Public License     *
*   along with this program.  If not, see <http://www.gnu.org/licenses/>. *
*************************************************************************]]

local Coordinates = {}

--- Converts global coordinates from amun to strategy local coordinates
-- @class function
-- @name toLocal
-- @param data Vector/number - vector or angle to convert
-- @return Vector/number

--[[
separator for luadoc]]--

--- Converts strategy local coordinates to global coordinates for amun
-- @class function
-- @name toGlobal
-- @param data Vector/number - vector or angle to convert
-- @return Vector/number

--[[
separator for luadoc]]--

--- Does toGlobal conversion for a list
-- @class function
-- @name listToGlobal
-- @param data (Vector/number)[] - list to map
-- @return (Vector/number)[]

--[[
separator for luadoc]]--

local function invertCoordinates(data)
	local dtype = type(data)
	if dtype == "number" then
		if data > math.pi then
			return data - math.pi
		else
			return data + math.pi
		end
	elseif dtype == "nil" then
		error("nil isn't a coordinate")
	else
		return Vector(-data.x, -data.y, data:isReadonly())
	end
end

local function invertList(data)
	local inverted = {}
	for k,v in ipairs(data) do
		inverted[k] = invertCoordinates(v)
	end
	return inverted
end

local function passthrough(data)
	return data
end

local function blueToVision(data)
	local dtype = type(data)
	if dtype == "number" then
		-- Rotate 270 degrees
		data = data - 1.5 * math.pi
		if data < 0 then
			data = data + 2 * math.pi
		end
		return data
	elseif dtype == "nil" then
		error("nil isn't a coordinate")
	else
		data = 1000 * data
		return Vector(-data.y, data.x, data:isReadonly())
	end
end

local function yellowToVision(data)
	local dtype = type(data)
	if dtype == "number" then
		-- Rotate 90 degrees
		data = data - 0.5 * math.pi
		if data < 0 then
			data = data + 2 * math.pi
		end
		return data
	elseif dtype == "nil" then
		error("nil isn't a coordinate")
	else
		data = 1000 * data
		return Vector(data.y, -data.x, data:isReadonly())
	end
end

function Coordinates._setIsBlue(teamIsBlue)
	if teamIsBlue then
		Coordinates.toGlobal = invertCoordinates
		Coordinates.toLocal = invertCoordinates
		Coordinates.listToGlobal = invertList
		Coordinates.toVision = blueToVision
	else
		Coordinates.toGlobal = passthrough
		Coordinates.toLocal = passthrough
		Coordinates.listToGlobal = passthrough
		Coordinates.toVision = yellowToVision
	end
end

return Coordinates
